src/azul/service/avro_pfb.py

```diff
@@ -0,0 +1,462 @@
+from operator import (
+    attrgetter,
+)
+from typing import (
+    Iterable,
+    MutableMapping,
+    MutableSet,
+    cast,
+)
+
+import attr
+import fastavro
+from fastavro.validation import (
+    ValidationError,
+)
+from more_itertools import (
+    one,
+)
+
+from azul import (
+    config,
+)
+from azul.indexer.document import (
+    FieldTypes,
+    null_bool,
+    null_float,
+    null_int,
+    null_int_sum_sort,
+    null_str,
+    pass_thru_int,
+    pass_thru_json,
+)
+from azul.plugins.metadata.hca.transform import (
+    pass_thru_uuid4,
+    value_and_unit,
+)
+from azul.types import (
+    JSON,
+    MutableJSON,
+)
+
+log = logging.getLogger(__name__)
+
+
+def write_pfb_entities(entities: Iterable[JSON], pfb_schema: JSON, path: str):
+    # fastavro doesn't know about our JSON type, hence the cast
+    parsed_schema = fastavro.parse_schema(cast(dict, pfb_schema))
+    with open(path, 'w+b') as fh:
+        # Writing the entities one at a time is ~2.5 slower, but makes it clear
+        # which entities fail, which is useful for debugging.
+        if config.debug:
+            log.info('Writing PFB entities individually')
+            for entity in entities:
+                try:
+                    fastavro.writer(fh, parsed_schema, [entity], validator=True)
+                except ValidationError:
+                    log.error('Failed to write Avro entity: %r', entity)
+                    raise
+        else:
+            fastavro.writer(fh, parsed_schema, entities, validator=True)
+
+
+class PFBConverter:
+    """
+    Converts documents from Elasticsearch into PFB entities. A document's inner
+    entities correspond to PFB entities which are normalized and linked via
+    Relations.
+    """
+
+    def __init__(self, schema: JSON):
+        self.schema = schema
+        self._entities: MutableMapping[PFBEntity,
+                                       MutableSet[PFBRelation]] = defaultdict(set)
+
+    def add_doc(self, doc: JSON):
+        """
+        Add an Elasticsearch document to be transformed.
+        """
+        contents = doc['contents']
+        file_relations = set()
+        for entity_type, entity in contents.items():
+            if entity_type != 'files':
+                if len(entity) == 0:
+                    pass
+                elif len(entity) == 1:
+                    entity = one(entity)
+                    if 'document_id' in entity:
+                        entity = PFBEntity.from_json(name=entity_type,
+                                                     object_=entity,
+                                                     schema=self.schema)
+                        if entity not in self._entities:
+                            self._entities[entity] = set()
+                        file_relations.add(PFBRelation.to_entity(entity))
+                    else:
+                        # FIXME: Protocol entities lack document ID so we skip
+                        #        for now
```

Comment:
Ugly. Please stick with with what I had.

Common sense should imply that C guide doesn't apply to heavily indented comments.

Reply:
Sure thing, I'll change it back.

I appreciate the contributing guide because I don't have to guess what you prefer. Since our opinions on style (and common sense) may differ, it's valuable to me to have a definitive source. 

Adding this ad-hoc exception to the C guide makes this confusing for me and brings the guessing back. If this is a new rule about indentation, I'd like to formalize it.

Reply:
If you can keep it simple, go right ahead.